subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/plugins/IvyPublishPlugin.java

```diff
@@ -156,10 +169,49 @@ private void createGenerateMetadataTask(final TaskContainer tasks, final IvyPubl
             generateTask.getPublication().set(publication);
             generateTask.getPublications().set(publications);
             generateTask.getOutputFile().convention(buildDir.file("publications/" + publicationName + "/module.json"));
+            disableGradleMetadataGenerationIfCustomLayout(repositories, generateTask);
         });
         publication.setModuleDescriptorGenerator(generatorTask);
     }
 
+    private void disableGradleMetadataGenerationIfCustomLayout(NamedDomainObjectList<IvyArtifactRepository> repositories, GenerateModuleMetadata generateTask) {
+        AtomicBoolean didWarn = new AtomicBoolean();
+        Spec<? super Task> checkStandardLayout = task -> {
+            boolean standard = repositories.stream().allMatch(this::hasStandardPattern);
+            if (!standard && !didWarn.getAndSet(true)) {
+                LOGGER.warn("Publication of Gradle Module Metadata is disabled because you have configured an Ivy repository with a non-standard layout");
```

Comment:
This is worth a mention in the upgrade notes IMO

Reply:
Does it? It's broken today...

Reply:
What is broken is consumption. So not all users may have realized it. I think that having an indication of the change is valuable.

Reply:
I understand that but there's no actionable item here. The upgrade guide explains how to upgrade. Here there's no action to take: upgrade and you get a bugfix...

Reply:
Ah, you just made me realize that it could be an issue after all.
Say you have a regular remote Ivy repo to which you publish. But you also have a local one, with custom layout (say simpler layout) that you use for validating your publication artifacts.

With this change, the real publication will lose the GMM while it was perfectly valid.

Makes sense?

Reply:
It does, but you'd still get the warning. Or we need to change how GMM generation works for Ivy repos. And back to square one.

Reply:
Sorry, should have been clearer: my scenario makes it worth mentioning as a potential breaking change IMO.